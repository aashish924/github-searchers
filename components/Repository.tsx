import React  from 'react'
import styled from 'styled-components';

import RepositoryModel from '../shared/repository/repository';
import User from './User';
import ExternalLink from './ExternalLink';


const RepositoryContainer = styled.div`
    margin: 8px 4px;
    display: inline-flex;
    flex-direction: column;
    align-items: center;
`;

const LabelContainer = styled.span`
    margin: 4px 0;
    font-size: 12px;
    background-color: #008c99;
    padding: 4px;
    color: white;
    display: inline;
`;


export const Stats: React.FC<{title: string, value: number}> = ({title, value}) => {
    return <div style={{ margin: "8px" }}>
        <LabelContainer style={{ fontWeight: 500 }}>{title}</LabelContainer>
        <LabelContainer style={{ backgroundColor: "#70ccd3" }}>{value}</LabelContainer>
    </div>;
};


export const Repository: React.FC<{data: RepositoryModel}> = (props) => {
  const repo = props.data;

  return <RepositoryContainer>
      <span style={{ margin: "4px", fontSize: "18px", fontWeight: "bold" }}>
          <ExternalLink url={repo.repositoryUrl} label={repo.name} />
      </span>
      <div style={{ display: "flex", flexWrap: "wrap" }}>
          <Stats title="Stars" value={repo.stars} />
          <Stats title="Forks" value={repo.forksCount} />
          <Stats title="Watchers" value={repo.watchers} />
          <Stats title="Open Issues" value={repo.openIssuesCount} />
      </div>
      <User data={repo.author} imgSize="small" />
  </RepositoryContainer>
};


export default Repository;