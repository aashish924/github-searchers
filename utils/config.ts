export const MINIMUM_SEARCH_TERM_LENGTH = 3;