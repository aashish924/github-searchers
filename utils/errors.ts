export class ApiError extends Error {
  static statusCode = 500;

  constructor(public message: string) {
    super(message);
    this.name = "ApiError";
    this.stack = (<any> new Error()).stack;
  }

  get status(): number {
    // @ts-ignore
    return this.constructor.statusCode;
  }
}


export class MethodNotAllowed extends ApiError {
  static statusCode = 405;

  constructor(public message: string) {
    super(message);
    this.name = "ApiError";
    this.stack = (<any> new Error()).stack;
  }
}


export class InvalidPayload extends ApiError {
  static statusCode = 400;

  constructor(public message: string) {
    super(message);
    this.name = "ApiError";
    this.stack = (<any> new Error()).stack;
  }
}