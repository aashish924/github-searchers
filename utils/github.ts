import User from '../shared/user/user';
import Repository from '../shared/repository/repository';
import sampleUsersData from './users';
import sampleRepoData from './repo';


export type GithubEntityType = 'users' | 'repositories';


export interface MetaInfo {
  total: number;
  previousPage: number|null;
  nextPage: number|null;
  totalPages: number;
}

interface GenericApiResponse {
  response: any;
  previousPage: number|null;
  nextPage: number|null;
  totalPages: number;
}

export interface SearchResult<T> {
  meta: MetaInfo;
  items: Array<T>;
}


abstract class BaseGithubClient {
  /*
   * Search for given items using Github search API
   *
   * @param type  github object type
   * @param text  search keyword
   */
  abstract _search: (type: GithubEntityType, text: string, page: number) => Promise<GenericApiResponse>;

  findRepositories = async (text: string, page: number=1): Promise<SearchResult<Repository>> => {
    const resp = await this._search('repositories', text, page);
    const response = resp.response;
    return {
      meta: {
        total: response.total_count,
        previousPage: resp.previousPage,
        nextPage: resp.nextPage,
        totalPages: resp.totalPages
      },
      items: response.body.items.map(item => (<Repository>{
        name: item.name,
        repositoryUrl: item.html_url,
        createdAt: item.created_at,
        forksCount: item.forks_count,
        openIssuesCount: item.open_issues_count,
        watchers: item.watchers,
        stars: item.stargazers_count,
        author: {
          name: item.author.login,
          avatarUrl: item.author.avatar_url,
          url: item.author.html_url
        }
      }))
    }
  };

  findUsers = async (text: string, page: number=1): Promise<SearchResult<User>> => {
    const resp = await this._search('repositories', text, page);
    const response = resp.response;
    return {
      meta: {
        total: response.total_count,
        previousPage: resp.previousPage,
        nextPage: resp.nextPage,
        totalPages: resp.totalPages
      },
      items: response.body.items.map(item => (<User>{
        name: item.login,
        avatarUrl: item.avatar_url,
        url: item.html_url
      }))
    }
  }
}


export class FakeGithubClient extends BaseGithubClient {

  // @ts-ignore
  _search = async (type: GithubEntityType, text: string, page: number): Promise<GenericApiResponse> => {
    const body = type === 'users' ? sampleUsersData : sampleRepoData;
    const totalPages = 3;
    return {
      response: body,
      previousPage: page > 1 ? (page - 1) : null,
      nextPage: page < totalPages ? (page + 1): null,
      totalPages: totalPages
    }
  };
}

