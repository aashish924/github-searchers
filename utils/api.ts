import {NextApiRequest, NextApiResponse} from "next";
import {ApiError, MethodNotAllowed} from "./errors";

export type HTTP_METHOD = "HEAD" | "GET" | "POST" | "PUT" | "DELETE";


const handleError = (res, err) => {
  if (err instanceof ApiError) {
    res.status(err.status).json({detail: err.message});
  } else {
    res.status(500).json({detail: err.toString()})
  }
};


export const apiResponse = (allowedMethods: HTTP_METHOD[], validator?: (payload: any) => void) => {

  const wrapper = (handler) => {
    return (req: NextApiRequest, res: NextApiResponse) => {
      try {
        // Check http method
        if(!req.method || allowedMethods.indexOf(<HTTP_METHOD>req.method) < 0) {
          throw new MethodNotAllowed(`Allowed methods: ${allowedMethods.join(', ')}`);
        }
        // Validate request payload
        if (validator) {
          const payload = req.body;
          validator(payload);
        }
        // Execute
        handler(req, res).catch(err => {
          console.error("cannot handle request: " + err.toString());
          handleError(res, err);
        });
      } catch (err) {
        handleError(res, err);
      }
    }
  };

  return wrapper;
};


export default apiResponse;