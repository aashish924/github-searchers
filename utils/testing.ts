import { NextApiRequest, NextApiResponse } from 'next';
import User from '../shared/user/user';
import Repository from '../shared/repository/repository';
import { SearchResult } from './backend/github';
import { cache } from './backend/api';


export const mockRequest = ({method, json, headers}: {method?: string, json?: any, headers?: any}) => {
  return {
    method: method || 'GET',
    body: json || {},
    headers: headers || {}
  } as NextApiRequest
};


export const mockResponse = () => {
  const res = { } as NextApiResponse;
  res.status = jest.fn().mockReturnValue(res);
  res.json = jest.fn().mockReturnValue(res);
  return res;
};


export const mockUser = ({name, avatarUrl, url}: {name?: string, avatarUrl?: string, url?: string}): User => {
  const user: User = {
    name: name || 'John Doe',
    url: url || 'https://github.com/benzid-wael/',
    avatarUrl: avatarUrl || 'https://avatars0.githubusercontent.com/u/4288931?s=60&v=4'
  };
  return user;
};


export const mockRepository = (): Repository => {
  return {
    name: 'Github Searchers',
    repositoryUrl: 'https://github.com/benzid-wael/github-searchers',
    author: {
      name: 'John Doe',
      url: 'https://github.com/benzid-wael/',
      avatarUrl: 'https://avatars0.githubusercontent.com/u/4288931?s=60&v=4'
    },
  } as Repository;
};


export const mockUserSearchResult = (): SearchResult<User> => {
  return {
    meta: {
      total: 1,
      prev: null,
      next: null,
      last: 1
    },
    items: [
        mockUser({})
    ]
  }
};


export const mockRepositorySearchResult = (): SearchResult<Repository> => {
  return {
    meta: {
      total: 1,
      prev: null,
      next: null,
      last: 1
    },
    items: [
        mockRepository()
    ]
  }
};


export class FakeJsonLogger {

  // @ts-ignore
  @cache({ prefix: 'test', version: 1 })
  log(message: string, error?: string) {
    if (error) {
      throw new Error(error);
    }

    return {message: message}
  }

}