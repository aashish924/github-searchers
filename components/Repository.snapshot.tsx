import * as React from 'react';
import renderer from 'react-test-renderer';

import Repository from './Repository';
import { mockRepository } from '../utils/testing';

it('Repository should render correctly', () => {
    const tree = renderer.create(<Repository data={mockRepository()} />).toJSON();
    expect(tree).toMatchSnapshot();
});
