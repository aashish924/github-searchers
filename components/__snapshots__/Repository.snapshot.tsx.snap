// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository should render correctly 1`] = `
<div
  className="sc-AxiKw iTIBYz"
>
  <span
    style={
      Object {
        "fontSize": "18px",
        "fontWeight": "bold",
        "margin": "4px",
      }
    }
  >
    <a
      className="sc-AxjAm lnbIcq"
      href="https://github.com/benzid-wael/github-searchers"
      target="_blank"
    >
      Github Searchers
    </a>
  </span>
  <div
    style={
      Object {
        "display": "flex",
        "flexWrap": "wrap",
      }
    }
  >
    <div
      style={
        Object {
          "margin": "8px",
        }
      }
    >
      <span
        className="sc-AxhCb bRNTTE"
        style={
          Object {
            "fontWeight": 500,
          }
        }
      >
        Stars
      </span>
      <span
        className="sc-AxhCb bRNTTE"
        style={
          Object {
            "backgroundColor": "#70ccd3",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "margin": "8px",
        }
      }
    >
      <span
        className="sc-AxhCb bRNTTE"
        style={
          Object {
            "fontWeight": 500,
          }
        }
      >
        Forks
      </span>
      <span
        className="sc-AxhCb bRNTTE"
        style={
          Object {
            "backgroundColor": "#70ccd3",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "margin": "8px",
        }
      }
    >
      <span
        className="sc-AxhCb bRNTTE"
        style={
          Object {
            "fontWeight": 500,
          }
        }
      >
        Watchers
      </span>
      <span
        className="sc-AxhCb bRNTTE"
        style={
          Object {
            "backgroundColor": "#70ccd3",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "margin": "8px",
        }
      }
    >
      <span
        className="sc-AxhCb bRNTTE"
        style={
          Object {
            "fontWeight": 500,
          }
        }
      >
        Open Issues
      </span>
      <span
        className="sc-AxhCb bRNTTE"
        style={
          Object {
            "backgroundColor": "#70ccd3",
          }
        }
      />
    </div>
  </div>
  <div
    className="sc-AxirZ kEGBWq"
  >
    <img
      src="https://avatars0.githubusercontent.com/u/4288931?s=60&v=4"
      style={
        Object {
          "borderRadius": "50%",
          "height": "72px",
          "margin": "4px",
          "width": "72px",
        }
      }
    />
    <a
      className="sc-AxjAm lnbIcq"
      href="https://github.com/benzid-wael/"
      style={
        Object {
          "fontSize": "18px",
        }
      }
      target="_blank"
    >
      John Doe
    </a>
  </div>
</div>
`;
